packages/core/src/electron-main/electron-main-menu.ts

```diff
@@ -0,0 +1,238 @@
+// *****************************************************************************
+// Copyright (C) 2022 Arduino SA and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { injectable } from 'inversify';
+import { app, BrowserWindow, ipcMain, WebContents } from '../../electron-shared/electron';
+import { Menu } from '../../electron-shared/electron/index';
+import { isOSX } from '../common';
+import { MenuItemConstructorOptions } from '../electron-common/menu';
+import {
+    CloseContextMenu,
+    ContextMenuDidClose, MenuItemDidClick, SetMenu, ShowContextMenu, UpdateMenuItems
+} from '../electron-common/messaging/electron-messages';
+import { ElectronMainApplicationContribution } from './electron-main-application';
+
+@injectable()
+export class ElectronMainMenu implements ElectronMainApplicationContribution {
+
+    // keys are the sender IDS, values are the current menu per window.
+    protected readonly menus = new Map<number, Menu | null>();
+    // keys are sender IDs, values are the open context menus per sender.
+    protected readonly openContextMenus = new Map<number, Map<string, Menu>>();
+
+    onStart(): void {
+        ipcMain.on(SetMenu.Signal, ({ sender }, params: SetMenu.Params) => this.setMenu({ sender }, params));
+        ipcMain.on(UpdateMenuItems.Signal, ({ sender }, params: UpdateMenuItems.Params) => this.updateMenuItems({ sender }, params));
+        ipcMain.on(ShowContextMenu.Signal, ({ sender }, params: ShowContextMenu.Params) => this.showContextMenu({ sender }, params));
+        ipcMain.on(CloseContextMenu.Signal, ({ sender }, params: CloseContextMenu.Params) => this.closeContextMenu({ sender }, params));
```

Comment:
Because the `sender` is sufficient in the implementation.

Reply:
Destructuring out the sender value is less performant than simply passing the pointer to the event (it requires both the destructuring itself and the allocation of the object containing only `{sender}` - trivial, but true) and limits the customization possible in the handlers, since they lack access to the rest of the fields on the event. Unless you foresee a use case where we need to perform these actions and _only_ have access to the `BrowserWindow`, it would be better - more extensible - to just pass around the event.